### ISSUE TITLE
Make requirements installation optional

### ISSUE THREAD

**Issue by tadejsv** [ultralytics/yolov5#3094]
## 🚀 Feature

I am using `yolov5` via `torch.hub`. When I load a model, the library attempts to install all the requirements. This is a problem for me, as the installed packages interfere with my system packages. Specifically, I have `opencv` installed via conda, but `yolo5` does not see it, and assumes it is missing, so it starts installing `opencv-python`. This then messes up some of `opencv` functionality. And in general, some packages that it tries to install are not strictly necessary - you only need PIL if you use it to load images, which I do not.

Would it be possible to make this installation optional, and just issue a warning if some packages are not found? This would make my life much easier :)

**Comment by glenn-jocher** [comment]
@tadejsv hi thanks for the feedback! Yes we are aware of this issue and are looking for a fix, see #3073 for a duplicate issue. Pillow is a requirement for Hub models as it is used throughout our autoshape wrapper.

Could you run the commands in https://github.com/ultralytics/yolov5/issues/3073#issuecomment-835788322 and report back your results? Thank you!

**Comment by tadejsv** [comment]
Hi @glenn-jocher - I ran the commands from the issue mentioned, both show no package found, even though `import cv2` works normally.

But I don't think that is a duplicate issue. My issue is not that in this particular case the right package is not picked up, it is that I would prefer that `yolo5` would not install any packages on its own, when loading it from torch hub. I am aware that Pillow is used for loading images - but only if you pass a filename (and not a numpy array), which is what I am not doing.

**Comment by glenn-jocher** [comment]
@tadejsv Pillow is a hard requirement. You can not run YOLOv5 PyTorch Hub models without it, regardless of your input method. If you're missing it and it is being auto-installed then everything is working as intended.

If your opencv installation is not found, then there is a bug and things are not working as intended.

I'm confused though, none of the 3 methods in https://github.com/ultralytics/yolov5/issues/3073#issuecomment-835788322 detected your `opencv-python` installation?

**Comment by tadejsv** [comment]
No, because it is not installed with pip. I am aware that things are "working as intended", that's why I opened this as a feature request and not a bug report :)

**Comment by tadejsv** [comment]
To clarify, the package I have installed is `py-opencv`, and not `opencv-python`. But it's also not detected by any of the 3 methods, though you can see it with `conda list`

**Comment by glenn-jocher** [comment]
@tadejsv you may not want to rely on 3rd party opencv packages, as their reliability is unknown, and also because they will cause conflicts with the requirements.txt dependencies.

One workaround if you would like to experiment with alternate packages is to simply fork this repo into your own directory, make any modifications you want to your forked hubconf.py, and then load YOLOv5 Hub models from your fork:
```python
import torch

# Model
model = torch.hub.load('your_github_account/yolov5', 'yolov5s')
```

You can comment the `check_requirements()` function here, or add `opencv-python` to the exclude list:
https://github.com/ultralytics/yolov5/blob/d6a71ae5f27c2b75c5a61846d5a756db534757ad/hubconf.py#L18

**Comment by tadejsv** [comment]
@glenn-jocher  I've had no issue with reliability so far. I am using the conda installation of opencv as it is compiled with some OpenCV modules which the pip installation does not have. I also prefer using conda as it makes sure all other system dependencies are installed.

I am aware I could fork the repo, I was just hoping for a more "elegant" solution that does not involve maintaining my own fork. Anyways, I see that changes to requirements installation here aren't something you are willing to consider, so I'll be closing this issue.

**Comment by glenn-jocher** [comment]
@tadejsv thanks for the feedback, we'll keep it in mind for future updates.

**Comment by glenn-jocher** [comment]
@tadejsv good news 😃! Your original issue may now be fixed ✅ in PR #3298. This PR increases robustness to failures in `check_requirements()` function during attempted package installs. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by tadejsv** [comment]
@glenn-jocher This does not solve my issue, as my issue was not with failures during installation, but rather with unwanted packages being installed

**Comment by glenn-jocher** [comment]
@tadejsv got it, thanks for the feedback!

**Comment by glenn-jocher** [comment]
@tadejsv good news 😃! Your original issue may now be fixed ✅ in PR #3495. This PR adds opencv-python to the hubconf.py check_requirements() `exclude` list, allowing for 3rd party or contrib versions of OpenCV to be used. To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
